filebeat/input/filestream/config.go

```diff
@@ -142,3 +144,78 @@ func (c *config) Validate() error {
 
 	return nil
 }
+
+// ValidateInputIDs checks all filestream inputs to ensure all have an ID and
+// the ID is unique. If there is any empty or duplicated ID, it logs an error
+// containing the offending input configurations and returns an error containing
+// the duplicated IDs.
+func ValidateInputIDs(inputs []*conf.C, logger *logp.Logger) error {
+	ids := make(map[string][]*conf.C)
+	var duplicates []string
+	var empty []*conf.C
+	for _, input := range inputs {
+		fsInput := struct {
+			ID   string `config:"id"`
+			Type string `config:"type"`
+		}{}
+		err := input.Unpack(&fsInput)
+		if err != nil {
+			return fmt.Errorf("failed to unpack filestream input configuration: %w", err)
+		}
+		if fsInput.Type == "filestream" {
+			if fsInput.ID == "" {
+				empty = append(empty, input)
+				continue
+			}
+			ids[fsInput.ID] = append(ids[fsInput.ID], input)
+			if len(ids[fsInput.ID]) > 1 {
+				duplicates = append(duplicates, fsInput.ID)
+			}
+		}
+	}
+
+	var errs []string
+	if empty != nil {
+		errs = append(errs, "input without ID")
+	}
+	if len(duplicates) != 0 {
+		errs = append(errs, fmt.Sprintf("filestream inputs with duplicated IDs: %v", strings.Join(duplicates, ",")))
+	}
+
+	if len(errs) != 0 {
+		jsonDupCfg := collectOffendingInputs(duplicates, empty, ids)
+		logger.Errorw("filestream inputs with invalid IDs", "inputs", jsonDupCfg)
+
+		return fmt.Errorf("filestream inputs validation error: %s", strings.Join(errs, ", "))
+	}
+
+	return nil
+}
+
+func collectOffendingInputs(duplicates []string, empty []*conf.C, ids map[string][]*conf.C) []map[string]interface{} {
+	var cfgs []map[string]interface{}
+
+	if len(empty) > 0 {
+		for _, cfg := range empty {
+			toJson := map[string]interface{}{}
+			err := cfg.Unpack(&toJson)
+			if err != nil {
+				toJson["emptyID"] = fmt.Sprintf("failed to unpack config: %v", err)
+			}
+			cfgs = append(cfgs, toJson)
+		}
+	}
+
+	for _, id := range duplicates {
+		for _, dupcfgs := range ids[id] {
+			toJson := map[string]interface{}{}
+			err := dupcfgs.Unpack(&toJson)
+			if err != nil {
+				toJson[id] = fmt.Sprintf("failed to unpack config: %v", err)
+			}
+			cfgs = append(cfgs, toJson)
+		}
+	}
```

Comment:
See my previous comment about `duplicates`.
Without the change above we would collect the same list of configs multiple times here, I don't think this is what's desired.